File: HW/HW12/HW12.java
private int[] parent;
private int[] rank;

public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
    parent = new int[n + 1];
    rank = new int[n + 1];

    for (int i = 0; i <= n; i++) {
        parent[i] = i;
        rank[i] = 0;
    }

    List<int[]> arr = new ArrayList<>();

    for (int i = 0; i < wells.length; i++) {
        arr.add(new int[]{wells[i], 0, i + 1});
    }

    for (int[] pipe : pipes) {
        arr.add(new int[]{pipe[2], pipe[0], pipe[1]});
    }

    Collections.sort(arr, Comparator.comparingInt(a -> a[0]));

    int result = 0;

    for (int[] connection : arr) {
        int cost = connection[0];
        int house1 = connection[1];
        int house2 = connection[2];
        if (union(house1, house2)) {
            result += cost;
        }
    }

    return result;
}

private int find(int x) {
    if (parent[x] != x) {
        parent[x] = find(parent[x]);
    }
    return parent[x];
}

private boolean union(int x, int y) {
    int rootX = find(x);
    int rootY = find(y);

    if (rootX == rootY) {
        return false;
    }

    if (rank[rootX] < rank[rootY]) {
        parent[rootX] = rootY;
    } else if (rank[rootX] > rank[rootY]) {
        parent[rootY] = rootX;
    } else {
        parent[rootY] = rootX;
        rank[rootX]++;
    }

    return true;
}

File: HW/HW12/HW12.py
def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
        parent = [i for i in range(n+1)]
        rank = [0]*(n+1)
        
        def find(house):
            if house != parent[house]:
                parent[house] = find(parent[house])
            return parent[house]
        
        def union(h1, h2):
            root1 = find(h1)
            root2 = find(h2)

            if root1 == root2:
                return False
            
            if rank[root1] > rank[root2]:
                parent[h2] = root1
            elif rank[root1] < rank[root2]:
                parent[h1] = root2
            else:
                parent[root1] = root2
                rank[root2] += 1
            
            return True

        arr = []
        for i in range(len(wells)):
            arr.append([wells[i], 0, i+1])
        
        for house1,house2,cost in pipes:
            arr.append([cost, house1, house2])
        
        arr.sort(key=lambda x: x[0])
        
        result = 0
        for cost,house1,house2 in arr:
            if union(house1, house2):
                result += cost
        return result
